model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProvider.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.models.map.storage.jpa;
+
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+import javax.persistence.EntityManager;
+import javax.sql.DataSource;
+import org.jboss.logging.Logger;
+import org.keycloak.Config;
+import org.keycloak.common.util.StringPropertyReplacer;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.dblock.DBLockManager;
+import org.keycloak.models.dblock.DBLockProvider;
+import org.keycloak.models.map.common.AbstractEntity;
+import org.keycloak.models.map.storage.MapStorage;
+import org.keycloak.models.map.storage.MapStorageProvider;
+import org.keycloak.models.map.storage.MapStorageProviderFactory.Flag;
+import org.keycloak.models.map.storage.jpa.updater.MapJpaUpdaterProvider;
+import org.keycloak.models.utils.KeycloakModelUtils;
+
+public class JpaMapStorageProvider implements MapStorageProvider {
+
+    private static final Logger logger = Logger.getLogger(JpaMapStorageProvider.class);
+    private final EntityManager em;
+    private final KeycloakSession session;
+    private final Config.Scope config;
+
+    public JpaMapStorageProvider(EntityManager em, KeycloakSession session, Config.Scope config) {
+        this.em = em;
+        this.session = session;
+        this.config = config;
+    }
+
+    @Override
+    public void close() {
+        em.close();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <V extends AbstractEntity, M> MapStorage<V, M> getStorage(Class<M> modelType, Flag... flags) {
+        init(modelType);
```

Comment:
Could the attempted updates be tracked in the factory to prevent locking / rechecking the update status upon every provider instantiation? Perhaps the whole `init` method logic could be put into the `JpaMapStorageProviderFactory`? This would enable removing reference to `config` in this class.